.github/workflows/on-pull-request-opened-and-synchronize.yml

```diff
@@ -77,35 +79,151 @@ jobs:
+      - name: Upload snapshots
+        uses: actions/upload-artifact@v4
+        with:
+          path: snapshots
+          name: visual-snapshots
+
+  test-visuals:
+    name: Test (Visuals)
+    # Invididual MacOS jobs are likely faster. But MacOS runner concurrency is
+    # limited to 5. So it would be overall slower and we'd soak up all of our
+    # account's MacOS runners.
+    runs-on: ubuntu-latest
+    # Increase sharding if the job takes longer than 5 minutes. We want it to
+    # be no slower than the `test` job, which is our slowest.
+    timeout-minutes: 5
+    needs:
+      - install-dependencies
+      - fetch-visual-snapshots
+    strategy:
+      # We need every job in the matrix to finish so we can produce a full report.
+      fail-fast: false
+      matrix:
+        shardIndex: [1, 2, 3, 4, 5]
+        shardTotal: [5]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/pnpm
+      - name: Download snapshots
+        uses: actions/download-artifact@v4
+        with:
+          path: dist/snapshots
+          name: visual-snapshots
+      - name: Test
+        # Failing would add quite a bit of time to CI. It would also add complexity, both
+        # for developers and to this workflow.
+        #
+        # To do it, we'd have to hack up GitHub's staged deployments feature to require an
+        # approval of the visual changes. After they're approved, we'd update the snapshots
+        # on R2. Then we'd to force a re-run of this job so it passes for the branch in question.
+        # Lastly, after merge, we'd have to synchronize the snapshots again in case another
+        # branch overwrote them in the meantime.
+        #
+        # So we simply expect the developer and reviewers to look over the test report, deployed
+        # below, before merging.
+        continue-on-error: true
+        run: npx playwright test --shard=${{ matrix.shardIndex }}/${{ matrix.shardTotal }}
+      - name: Upload blob report
+        if: ${{ !cancelled() }}
+        uses: actions/upload-artifact@v4
+        with:
+          name: visual-test-report-blob-${{ matrix.shardIndex }}
+          path: blob-report
+
+  deploy-visual-test-report:
+    name: Deploy Test Report
+    if: ${{ !cancelled() }}
+    runs-on: ubuntu-latest
+    needs:
+      - test-visuals
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./.github/actions/pnpm
+      - name: Download blob reports
+        uses: actions/download-artifact@v4
+        with:
+          merge-multiple: true
+          path: visual-test-report-blob
+          pattern: visual-test-report-blob-*
+      - name: Merge blob reports
+        run: |
+          PLAYWRIGHT_HTML_ATTACHMENTS_BASE_URL=https://${{ vars.CLOUDFLARE_CUSTOM_DOMAIN }}/${{ github.head_ref }}/visual-test-report/data/ npx playwright merge-reports --reporter html ./visual-test-report-blob
```

Comment:
A sharded run each has its own report blob. The blobs are then combined to produce the a full HTML report.